mlir/lib/Dialect/Tensor/IR/TensorDialect.cpp

```diff
@@ -8,8 +8,11 @@
 
 #include "mlir/Dialect/Affine/IR/AffineOps.h"
 #include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h"
 #include "mlir/Dialect/Complex/IR/Complex.h"
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Transform/IR/TransformInterfaces.h"
```

Comment:
I'm not sure this is desirable.

Reply:
Does [Include from Tensor instead of Transform](https://github.com/llvm/llvm-project/pull/78368/commits/895e56461f39ea78b70ad709c4434580181af402) address this issue? If not, could you expand on why this isn't desireable?

Reply:
The issue I see is that we don't want Tensor/IR, or any other dialect definition for that matter, to depend on the Transform dialect. The latter is a sort of utility that would belong to lib/IR or lib/Transforms, but for now matures as a separate dialect and should remain loosely tied to the rest of the infrastructure. So far, this was solved by having Tensor/TransformOps in a separate library that depends on both Transform and Tensor dialects. I'm a bit surprised that we are not running into a dependency cycle here, probably just cmake not modeling header-level dependencies, but other build systems will complain.

I see two temporary solutions:
1. Revert to depending on Transform here, at least not creating a dependency cycle. (I can see about factoring out the interfaces part of the transform into a separate library because we only ever those, not the rest of lib/IR.)
2. Consider putting these promises in the [`TensorTransformDialectExtension`](https://github.com/llvm/llvm-project/blob/d022f32c73c57b59a9121eba909f5034e89c628e/mlir/lib/Dialect/Tensor/TransformOps/TensorTransformOps.cpp#L222). Not entirely convinced this is the best place though as one may be needing these implementations even without the additional ops provided by the tensor transform extension.

Reply:
Fundamentally in-tree: adding a promise **requires** the dialect to know about the interface class.
That means if you want to have an interface for the Transform dialect, and implement it for other in-tree dialects, these dialects must be able to know about this interface to declare the promise.
So this:

> The issue I see is that we don't want Tensor/IR, or any other dialect definition for that matter, to depend on the Transform dialect.

Is an impossible requirement if "depend on the Transform dialect" implies "can't include the interface declaration from the Transform dialect".
The only option to fullfill your criteria is to entirely delete the Transform interface implementation for these dialects from the repo and consider that the Transform dialect can't implement interfaces for these dialects.  I don't think this is what you want though.

Reply:
I see a semi-satisfactory solution: isolate transform dialect interfaces into a separate library that does not depend on the dialect definition itself. Tensor dialect can then depend/include from that, and so can other dialects, without additional cost. Eventually, that interface library can move under `lib/Interfaces` as it will follow exactly the same model. (There is an adjacent discussion on `lib/Interfaces`, but I don't want to block progress here on that).